### ISSUE TITLE
Script Fails in JAMF using policy with script upload

### ISSUE THREAD

**Issue by nertles7** [grahampugh/erase-install#227]
**Describe the bug**
The script fails with script exit code 7 after about 14s. I am attempting to update an M1 MacBook Air with 11.6 to 12.3.

**To Reproduce**
I have uploaded your script into JAMF and have set a policy with parameters --force-curl and --reinstall and provided it as a Self Service item on my test machine.
 
**Expected behavior**
The mac should have upgraded to 12.3

**Code/log output**
JAMF Policy output
```
[STEP 1 of 4]
--
Executing Policy Update macOS
[STEP 2 of 4]
Running script erase-install for latest macOS...
Script exit code: 7
Script result:    
[erase-install] v26.1 script execution started: Wed Mar 30 15:26:11 EDT 2022    
[erase-install] Caffeinating this script (pid=2057)    
[check_free_space] OK - 64 GB free/purgeable disk space detected    
[erase-install] Looking for existing installer app or pkg    
[find_existing_installer] Installer sparse image found at /Library/Management/erase-install/Install_macOS_12.3-21E230.sparseimage. /dev/disk4          	GUID_partition_scheme          	 /dev/disk4s1        	EFI                            	 /dev/disk4s2        	Apple_HFS                      	/Volumes/Install_macOS_12.3-21E230    
[check_installer_is_valid] Checking validity of /Volumes/Install_macOS_12.3-21E230/Applications/Install macOS Monterey.app. Volume Shared Support on disk5s2 force-unmounted    
[check_installer_is_valid] Mounting /Volumes/Install_macOS_12.3-21E230/Applications/Install macOS Monterey.app/Contents/SharedSupport/SharedSupport.dmg   
 [check_installer_is_valid] Using Build value from com_apple_MobileAsset_MacSoftwareUpdate.xml Volume Shared Support on disk5s2 force-unmounted    
[compare_build_versions] Comparing (1) 20G165 with (2) 21E230   
 [check_installer_is_valid] Installer: 21E230 >= System: 20G165 : valid build. yes resstudent is a member of everyone    
[check_password] Success: the password entered is the correct login password for resstudent.    
[erase-install] Installer is at: /Volumes/Install_macOS_12.3-21E230/Applications/Install macOS Monterey.app    
[erase-install] Closing jamfHelper download message (language=en)     
[erase-install] WARNING! Running /Volumes/Install_macOS_12.3-21E230/Applications/Install macOS Monterey.app with reinstall option     
[erase-install] Opening jamfHelper message (language=en) Error: could not get authorization... By using the agreetolicense option, you are agreeing that you have run this tool with the license only option and have read and agreed to the terms. If you do not agree, press CTRL-C and cancel this process immediately.     
[erase-install] attempting to terminate the 'caffeinate' process - Termination message indicates success /Library/Application Support/JAMF/tmp/erase-install for latest macOS: line 1055:  2069 Terminated: 15          /usr/bin/caffeinate -dimsu -w $
Error running script: return code was 7.
[STEP 3 of 4]
[STEP 4 of 4]

```

**Environment (please complete the following information):**
 - 11.6
 - erase-install version 26.1 (Have also tried 26.0 and 25.0 the same error occurs.)

**Comment by grahampugh** [comment]
I suspect there's a problem with that downloaded installer. Use `--overwrite` to replace it and try again.

**Comment by nertles7** [comment]
I tried that, I also tried the option that does the full download from Apple, that didn't help either. I also tried a different network, sometimes our content filter Zscaler messes with things, but same issue on our network that just uses a fortigate with minimal blocking. I then DFUed the machine to 11.5.2 and the same policy no changes worked fine. I tried DFUing the machine back to 11.6 and it failed again, same error. I noticed on the logs this line 
```
[erase-install] Opening jamfHelper message (language=en) Error: could not get authorization... By using the agreetolicense option, you are agreeing that you have run this tool with the license only option and have read and agreed to the terms. If you do not agree, press CTRL-C and cancel this process immediately.  
```
Where it says could not get authorization was not present on the 11.5.2 install. I am DFUing to 11.3.1 and going to try it on that version of macOS. I may also pull another MacBook Air and try it on 11.6 to confirm tomorrow.

**Comment by FrageDev** [comment]
"Error: could not get authorization..."

Same error I reported, this is a Secure Token Issue. Check who have the Secure Token Enabled and if this user is used to start the startosinstall.

Also if this account is an admin account created by Jamf you might have so issue saying the user is not admin because Jamf is using dscl which is not giving full admin like the sysadminctl command.

**Comment by grahampugh** [comment]
This script checks that the user has a Secure Token. It also elevates a user to admin for the --erase option. 

However, it won't do that if it thinks the user is already an admin. That's good information about Jamf using dscl. Do you know a way to check if the user is not a "proper" admin?

**Comment by ryangball** [comment]
I'd be surprised to know that the Jamf binary would use something like dscl rather than OpenDirectory or another API to create users. It not creating accounts with SecureTokens doesn't reveal their account creation method.

**Comment by FrageDev** [comment]
> This script checks that the user has a Secure Token. It also elevates a user to admin for the --erase option.
> 
> However, it won't do that if it thinks the user is already an admin. That's good information about Jamf using dscl. Do you know a way to check if the user is not a "proper" admin?

Jamf using dscl is hypothetic, I'm not sure but all I know from my issue and testing is admin created with Jamf and also dscl doesn't give a real admin state compare to a sysadminctl -admin. Even if you compare group etc all match. Might be a macOS bug

**Comment by grahampugh** [comment]
@nertles7 does your restrictions profile prevent access to Disk Images? The default method of this script requires the ability to mount a disk image (sparseimage). You might be able to get around that using the `--pkg` or `--fetch-full-installer` options, as I don't think either of these options need to mount a Disk Image.

**Comment by nertles7** [comment]
No I actually see the disk image mount fine. I have tried the fetch full installer as well to no avail. After my testing the following circumstances lead to a failure:

1) macOS 11.6
2) Our student lock down profile is applied at the time of enrollment.

The same lockdown profile can be on a Mac on 11.5.2 and 11.3.1 and it will work fine. I have also tested an 11.6 Mac that got the student profile at the time of enrollment, then taking the student lockdown off, it will still fail. If the Mac is enrolled without the student profile on, but is put on later it works. So it would seem to me something about 11.6 and our student lockdown profile being pushed at the time of enrollment taints the Mac and causes this error. The user I am using has secure token and volume ownership, of course your script verifies that anyway.

What exact setting in our lockdown profile would probably be much harder to find, we have quite a lot blocked.
